model/common/src/icon4py/model/common/grid/grid_refinement.py

```diff
@@ -0,0 +1,183 @@
+# ICON4Py - ICON inspired code in Python and GT4Py
+#
+# Copyright (c) 2022-2024, ETH Zurich and MeteoSwiss
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+import logging
+from types import ModuleType
+from typing import Final
+
+import numpy as np
+from gt4py import next as gtx
+
+from icon4py.model.common import dimension as dims
+from icon4py.model.common.grid import horizontal as h_grid
+from icon4py.model.common.utils import data_allocation as data_alloc
+
+
+"""
+Refinement control for ICON grid.
+
+Grid refinement is used in the context of
+- local area grids to determine the type of a grid point: its distance from the boundary
+- nested horizontal grids to determine the nested overlap regions for feedback
+- domain decomposition to order grid points
+
+See Zaengl et al. Grid Refinement in ICON v2.6.4 (Geosci. Model Dev., 15, 7153-7176, 202)
+
+This module only contains functionality related to grid refinement as we use it in ICON4Py.
+
+"""
+_log = logging.getLogger(__name__)
+
+
+_MAX_BOUNDARY_DISTANCE: Final[dict[gtx.Dimension, int]] = {
+    dims.CellDim: 14,
+    dims.EdgeDim: 28,
+    dims.VertexDim: 14,
+}
+"""
+Grid points in the grid refinement fields are labeled with their distance to the lateral boundary.
+The distance is counted in "rows" and distances _MAX_BOUNDARY_DISTANCE are computed.
+
+**Rows** can be defined like:
+### CELLS:
+1. row: describes all cells that share *at least* 1 vertex with the lateral boundary
+2. row: shares at least 1 vertex with the 1. row
+
+### VERTEX:
+1. row: vertex on the lateral boundary
+2. row: shared between cell of 1. row and 2. row cell.
+...
+
+## EDGES:
+1. row: edge on boundary
+2. row: edge only between cells of 1. row
+3. row: edge between cells of 1. row and 2. row
+4. row: edges only between 2. row cells
+...
+
+
+Grid points can be ordered (moved to the beginning of an field array) up to any distance <= MAX_BOUNDARY_DISTANCE in the grid-generators, the exact number of
+ordered rows is a parameter to the grid generation.
+"""
+
+_GRID_REFINEMENT_BOUNDARY_WIDTH: Final[dict[gtx.Dimension, int]] = {
+    dims.CellDim: 4,
+    dims.EdgeDim: 9,
+    dims.VertexDim: 4,
+}
+"""Number of cell rows used in ICON grid refinement: meaning there are for lateral boundary cell-rows.
+"""
+
+_UNORDERED: Final[dict[gtx.Dimension, tuple[int, int]]] = {
+    dims.CellDim: (0, -4),
+    dims.EdgeDim: (0, -8),
+    dims.VertexDim: (0, -4),
+}
+"""Refinement value indicating a point is in the unordered interior (fully prognostic) of the grid: this is encoded by 0 or -4 in coarser parent grid."""
```

Comment:
It mean the tuple is to be read as: either `0` or `-4`? (not range or something else)